ggml.c

```diff
@@ -11502,6 +11513,16 @@ static void ggml_compute_forward_soft_max_f32(
             ggml_vec_acc_f32(nc, wp, mp);
         }
 
+        // alibi bias
+        if (src2) {
+            const int   h     = (i1/ne01)%ne02;
+            const float slope = ((float *)(src2->data))[h];
+
+            for (int i = 0; i < nc; i++) {
+                wp[i] = wp[i] + slope*i;
```

Comment:
Currently, the `test-backend-ops` are failing on the CPU in a very weird way:

```bash
make -j && ./bin/test-backend-ops -o SOFT_MAX -b CPU
```

```
...
  SOFT_MAX(type=f32,ne=[1,36,1,1],mask=0,slope=0,scale=0.100000): OK
  SOFT_MAX(type=f32,ne=[2,4,1,1],mask=1,slope=1,scale=0.100000): OK
  SOFT_MAX(type=f32,ne=[4,20,1,1],mask=1,slope=1,scale=0.100000): OK
  SOFT_MAX(type=f32,ne=[3,19,1,1],mask=0,slope=1,scale=0.100000): [SOFT_MAX] NMSE = 0.228123672 > 0.000000100 FAIL
  SOFT_MAX(type=f32,ne=[4,5,1,1],mask=0,slope=0,scale=0.100000): OK
  SOFT_MAX(type=f32,ne=[4,23,1,1],mask=0,slope=0,scale=0.100000): OK
  SOFT_MAX(type=f32,ne=[2,3,1,1],mask=0,slope=0,scale=0.100000): OK
  SOFT_MAX(type=f32,ne=[2,22,1,1],mask=0,slope=0,scale=0.100000): OK
  SOFT_MAX(type=f32,ne=[3,43,1,1],mask=0,slope=0,scale=0.100000): OK
  SOFT_MAX(type=f32,ne=[2,16,1,1],mask=0,slope=0,scale=0.100000): OK
  SOFT_MAX(type=f32,ne=[4,35,1,1],mask=0,slope=0,scale=0.100000): OK
  SOFT_MAX(type=f32,ne=[8,28,1,1],mask=1,slope=1,scale=0.100000): OK
  SOFT_MAX(type=f32,ne=[4,28,1,1],mask=1,slope=1,scale=0.100000): OK
  SOFT_MAX(type=f32,ne=[8,3,1,1],mask=0,slope=1,scale=0.100000): [SOFT_MAX] NMSE = 0.710118388 > 0.000000100 FAIL
  SOFT_MAX(type=f32,ne=[7,27,1,1],mask=0,slope=0,scale=0.100000): OK
  SOFT_MAX(type=f32,ne=[6,13,1,1],mask=0,slope=0,scale=0.100000): OK
  SOFT_MAX(type=f32,ne=[4,39,1,1],mask=0,slope=0,scale=0.100000): OK
...
```

But if I change this line like this:

```suggestion
                wp[i] = wp[i] + slope;
```

The tests on the CPU now all pass. Any ideas?
I feel I've made some stupid mistake, but this is quite unexpected observation and I can't figure out where is the problem.

I think the implementation is correct, because MPT generations match `master`

Reply:
The CPU backend test compares two identical runs on the CPU backend, so it there are differences that must be because the op produces different results for the same inputs, which may indicate out of bound accesses or synchronization issues. It could also be a bug in `test-backend-ops` or the graph copying mechanism, but that would be weird.

Reply:
I think I see the issue. The `src` are in many places only checked up to the first `NULL` src. So if the `mask` is `NULL`, it will skip copying the `slope`.

Reply:
Ah, that makes sense!

Reply:
This fixes it:

```diff
diff --git a/ggml-alloc.c b/ggml-alloc.c
index c28c37c4..d4123564 100644
--- a/ggml-alloc.c
+++ b/ggml-alloc.c
@@ -551,7 +551,7 @@ static void ggml_gallocr_alloc_graph_impl(ggml_gallocr_t galloc, struct ggml_cgr
         }
         for (int j = 0; j < GGML_MAX_SRC; j++) {
             if (graph->nodes[i]->src[j] == NULL) {
-                break;
+                continue;
             }
             if (graph->nodes[i]->src[j]->flags & GGML_TENSOR_FLAG_INPUT) {
                 ggml_gallocr_allocate_node(galloc, graph->nodes[i]->src[j], get_node_buffer_id(node_buffer_ids, i));
@@ -787,7 +787,7 @@ static bool ggml_gallocr_needs_realloc(ggml_gallocr_t galloc, struct ggml_cgraph
         for (int j = 0; j < GGML_MAX_SRC; j++) {
             struct ggml_tensor * src = node->src[j];
             if (src == NULL) {
-                break;
+                continue;
             }
             if (!ggml_gallocr_node_needs_realloc(galloc, src, node_alloc, &node_alloc->src[j])) {
 #ifndef NDEBUG
@@ -833,7 +833,7 @@ bool ggml_gallocr_alloc_graph(ggml_gallocr_t galloc, struct ggml_cgraph * graph)
         for (int j = 0; j < GGML_MAX_SRC; j++) {
             struct ggml_tensor * src = node->src[j];
             if (src == NULL) {
-                break;
+                continue;
             }
             ggml_gallocr_init_tensor(galloc, src, node_alloc, &node_alloc->src[j]);
         }
diff --git a/ggml-backend.c b/ggml-backend.c
index 87eea844..41f413ca 100644
--- a/ggml-backend.c
+++ b/ggml-backend.c
@@ -1040,7 +1040,7 @@ static int ggml_backend_sched_backend_id_from_cur(ggml_backend_sched_t sched, st
     for (int i = 0; i < GGML_MAX_SRC; i++) {
         const struct ggml_tensor * src = tensor->src[i];
         if (src == NULL) {
-            break;
+            continue;
         }
         if (src->buffer != NULL && src->buffer->usage == GGML_BACKEND_BUFFER_USAGE_WEIGHTS) {
             int src_backend = ggml_backend_sched_backend_from_buffer(sched, src->buffer);
@@ -1087,7 +1087,7 @@ static void ggml_backend_sched_print_assignments(ggml_backend_sched_t sched, str
         for (int j = 0; j < GGML_MAX_SRC; j++) {
             struct ggml_tensor * src = node->src[j];
             if (src == NULL) {
-                break;
+                continue;
             }
             ggml_backend_t src_backend = tensor_backend(src);
             fprintf(stderr, " %20.20s (%5.5s) [%5.5s %8.8s]", src->name,
@@ -1143,7 +1143,7 @@ static void ggml_backend_sched_split_graph(ggml_backend_sched_t sched, struct gg
         for (int j = 0; j < GGML_MAX_SRC; j++) {
             struct ggml_tensor * src = node->src[j];
             if (src == NULL) {
-                break;
+                continue;
             }
             if (tensor_backend_id(src) == -1) {
                 tensor_backend_id(src) = ggml_backend_sched_backend_id_from_cur(sched, src);
@@ -1255,7 +1255,7 @@ static void ggml_backend_sched_split_graph(ggml_backend_sched_t sched, struct gg
         for (int j = 0; j < GGML_MAX_SRC; j++) {
             struct ggml_tensor * src = node->src[j];
             if (src == NULL) {
-                break;
+                continue;
             }
             int src_backend_id = tensor_backend_id(src);
             if (src_backend_id == -1) {
@@ -1314,7 +1314,7 @@ static void ggml_backend_sched_split_graph(ggml_backend_sched_t sched, struct gg
             for (int j = 0; j < GGML_MAX_SRC; j++) {
                 struct ggml_tensor * src = node->src[j];
                 if (src == NULL) {
-                    break;
+                    continue;
                 }
                 int src_backend_id = tensor_backend_id(src);
                 assert(src_backend_id != -1); // all inputs should be assigned by now
@@ -1361,7 +1361,7 @@ static void ggml_backend_sched_split_graph(ggml_backend_sched_t sched, struct gg
         for (int j = 0; j < GGML_MAX_SRC; j++) {
             struct ggml_tensor * src = node->src[j];
             if (src == NULL) {
-                break;
+                continue;
             }
             ggml_backend_t src_backend = tensor_backend(src);
             if (src_backend != tensor_backend /* && src_backend != NULL */) {
@@ -1667,7 +1667,7 @@ static struct ggml_tensor * graph_copy_dup_tensor(struct ggml_hash_set hash_set,
     for (int i = 0; i < GGML_MAX_SRC; i++) {
         struct ggml_tensor * s = src->src[i];
         if (s == NULL) {
-            break;
+            continue;
         }
         dst->src[i] = graph_copy_dup_tensor(hash_set, node_copies, ctx_allocated, ctx_unallocated, s);
     }
@@ -1696,7 +1696,7 @@ static void graph_copy_init_tensor(struct ggml_hash_set hash_set, struct ggml_te
     for (int i = 0; i < GGML_MAX_SRC; i++) {
         struct ggml_tensor * s = src->src[i];
         if (s == NULL) {
-            break;
+            continue;
         }
         graph_copy_init_tensor(hash_set, node_copies, node_init, s);
     }
```

Does it make sense, or is there something better we can do?